File: java/567.permutation-in-string.java
/*
 * @lc app=leetcode id=567 lang=java
 *
 * [567] Permutation in String
 */

// @lc code=start
class Solution {
    public boolean checkInclusion(String s1, String s2) {
        if(s1.length() > s2.length()) return false;

        int[] freq = new int[26];

        for(int i = 0; i < s1.length(); i++) {
            freq[s1.charAt(i) - 'a']++;
        }

        for(int i = 0; i < s1.length(); i++) {
            freq[s2.charAt(i) - 'a']--;
        }

        boolean zero = true;
        for(int f : freq) {
            if(f != 0) {
                zero = false;
                break;
            }
        }

        if(zero) return true;

        for(int r = s1.length(); r < s2.length(); r++) {
            int l = r - s1.length();
            freq[s2.charAt(l) - 'a']++;
            freq[s2.charAt(r) - 'a']--;

            zero = true;
            for(int f : freq) {
                if(f != 0) {
                    zero = false;
                    break;
                }
            }

            if(zero) return true;
        }

        return false;
    }
}
// @lc code=end

File: cpp/567.permutation-in-string.cpp
/*
 * @lc app=leetcode id=567 lang=cpp
 *
 * [567] Permutation in String
 */

// @lc code=start
class Solution {
public:
    bool checkInclusion(string s1, string s2) {
        if(s1.size() > s2.size()) return false;

        int freq[26] = {};

        for(char c : s1) {
            freq[c-'a']++;
        }

        for(int r = 0; r < s1.size(); r++) {
            freq[s2[r]-'a']--;
        }

        int l = 0;

        for(int r = s1.size(); r < s2.size(); r++) {
            if(std::count(&freq[0], &freq[26], 0) == 26) {
                return true;
            }

            freq[s2[r]-'a']--;
            freq[s2[l]-'a']++;
            l++;
        }

        if(std::count(&freq[0], &freq[26], 0) == 26) {
            return true;
        }

        return false;
    }
};
// @lc code=end

